README.md
# Books-Search-Engine
Project for our Information Management exam, a vertical Search Engine about books


scraping/History.txt
History
Giuseppe Genna
narrativa italiana fantascienza
https://www.qlibri.it/narrativa-italiana/fantascienza/History/
Anno del Signore 2018: il mondo è trasformato. Il futuro è crollato nel presente, aggiornandolo e mutandolo. L'accelerazione tecnologica riconfigura tutto e tutti. Le macchine e gli algoritmi si candidano a mutare geneticamente il pianeta e l'umanità. Nulla e nessuno è indenne: il lavoro non è più lavoro, il denaro è puro fantasma, la specie umana è pronta a ibridarsi, persino la biologia rischia di non essere più biologica e la Storia non è più storia. Nella città più avanzata d'Italia si è installato un tecnopolo, in cui sta vedendo la luce un nuovo tipo di mente: un'intelligenza artificiale misteriosa e incomprensibile, a cui gli umani lavorano con dedizione cieca e speranze supreme. In questo bacino occidentale prospera allo stato bacillare il personaggio di uno scrittore, disoccupato e privo di qualunque riconoscimento, che riesce a trovare l'ultimo lavoro: interagirà, a vantaggio della mente artificiale, con una bambina altrettanto misteriosa, History, figlia di un tycoon della finanza, che soffre di una forma di autismo assoluto. L'intelligenza artificiale è molto interessata ai modi in cui History sente e reagisce alla realtà, vivendo in se stessa scene terrifiche e visioni infernali, dominate dalla presenza di una Trista Figura, una sorta di Uomo Nero che la invita alla scomparsa. In una deflagrazione di complotti e di sorprendenti svolte, il teatro umano che agisce in questo libro va incontro al momento decisivo nella storia della specie, entrando in un piano di realtà ulteriore, dove va in scena la verità di tutte le verità, un inedito horror della mente e dei corpi. Per raccontare il futuro che sta velocemente alterando il nostro presente, la scrittura metafisica di Genna intraprende una sfida all'ultimo sangue con la materia e con la lingua della narrazione estrema, rappresentando una tragedia classica in forma di autofiction e di profezia. A partire da un antefatto visionario, che consegna al lettore un intero tempo italiano trascorso e che vale un libro all'interno del libro, History è un lungo e vertiginoso precipitare verso una scena assoluta, in cui si assiste all'ultima trasformazione: quella dell'umano in una nuova forma rivista e corretta, non meno commovente e demonica della precedente, a cui noi tutti ancora apparteniamo, ogni giorno sempre meno.



indexing.py
import os
from whoosh.index import create_in
from whoosh.fields import *
import sys

import nltk
from nltk.corpus import stopwords
from nltk.tokenize import sent_tokenize, word_tokenize
 
def tokenize(text):
    # tokenizzazione
    tokens = nltk.word_tokenize(text)
    
    #Eliminazione di stopwords e punteggiatura
    tokens2=[]
    for t in tokens:
        if t not in stopwords.words('italian') and t.isalnum():
            tokens2.append(t)
    return tokens2
    

def createSearchableData(root):   
 
    ## definisco lo schema del mio indice
    schema = Schema( title=TEXT(stored=True),\
                     author=TEXT(stored=True),\
                     genre=KEYWORD(stored=True), \
                     link=ID(stored=True), \
                     path=ID(stored=True), \
                     content=TEXT(stored=True),\
                     contentData=TEXT)
                     
    ## creo la directory indexdir
    if not os.path.exists("indexdir"):
        os.mkdir("indexdir")
 
    ## Creo un indexWriter, che aggiunga i documenti secondo lo schema
    ix = create_in("indexdir",schema)
    writer = ix.writer()
 
    ## Trovo i file nella directory, e ne salvo i percorsi
    filepaths = [os.path.join(root,i) for i in os.listdir(root)]
    
    num = 1
    # per ogni percorso trovato...
    for path in filepaths:
        print(num)
        num+=1
    
        fp = open(path,'r')
        
        # Nella prima riga ho messo il titolo, nella seconda l'autore, nella terza il genere, nella quarta il link
        fileTitle = fp.readline()
        fileAuthor = fp.readline()
        fileGenre = fp.readline()
        fileLink = fp.readline()
        
        # Tutto il resto del file è occupato dalla trama
        filePlot = fp.read()
        
        # la sezione contentData è data dalle trame preprocessate
        fileData = tokenize(filePlot)
        
        ## Aggiungo un documento all'indice, con tutti i campi necessari
        writer.add_document( title = fileTitle,\
                             path = path,\
                             author = fileAuthor,\
                             genre = fileGenre,\
                             link = fileLink,\
                             content = filePlot,\
                             contentData = fileData)
        fp.close()
    writer.commit()
 
root = "scraping"
createSearchableData(root)


querying.py
from whoosh.index import open_dir
from whoosh.fields import *
from whoosh.qparser import QueryParser, MultifieldParser
import os, os.path
from whoosh import qparser

ix = open_dir("indexdir")

# Faccio digitare la query dall'utente
userQuery = input("Inserire una parola o frase da cercare: ")

searcher = ix.searcher()
#parser = QueryParser("contentData", schema=ix.schema)
#parser = qparser.QueryParser("contentData", schema=ix.schema, group=qparser.OrGroup)

#implementa la ricerca anche per titolo, autore o genere
parser = MultifieldParser(["contentData", "title", "author", "genre"], schema=ix.schema) 
query = parser.parse(userQuery)
results = searcher.search(query)
 

# Restituisco i primi topN risultati - se esistono
topN = int(10) # il 10 l'ho messo a caso
i=1

for r in results:
    if i > topN:
        break
    print(i, str(r.score), r['title'],r['author'], r['genre'])
    print(r['content'][:255]+"...\n\n")
    i+=1


qlibriScrape.py
from selenium import webdriver
from selenium.webdriver.firefox.service import Service
from selenium.webdriver.firefox.options import Options
from bs4 import BeautifulSoup
import pandas as pd

def replaced(str):
    # sostituisce alcuni caratteri di una stringa con altri.
    # serve per il passaggio da titolo/autore a parte del link
    str = str.replace(" ","-")
    str = str.replace("à","a")
    str = str.replace("è","e")
    str = str.replace("é","e")
    str = str.replace("ì","i")
    str = str.replace("ò","o")
    str = str.replace("ó","o")
    str = str.replace("ö","o")
    str = str.replace("ù","u")
    str = str.replace("ð","d")
    return str


## Funzione che fa effettivamente lo scraping, separata dal resto per leggibilità
def soupScraping(driver, titles, authors, plots):
    content = driver.page_source
    soup = BeautifulSoup(content, 'html.parser')

    ## Estraggo le info e le metto negli array appositi
    for a in soup.findAll(class_='listItem'):
        title=a.find('a',href=True, attrs={'class':'jr_listingTitle'})
        author=a.find('div', attrs={'class':'fieldValue'})
        plot=a.find('div', attrs={'class':'contentIntrotext'})
    
        # piccolo controllo, se non interrompo il ciclo mi dà errore
        if title is None or author is None or plot is None:
            break
        titles.append(title.text)
        authors.append(author.text)
        plots.append(plot.text) 
    # Ritorno gli array per aggiornarli
    return titles, authors, plots
    


### Roba per far andare via il warning e caricare la pagina in fretta
# options = Options()
# options.headless = False
# service = Service('/usr/lib/firefox/geckodriver')

# driver = webdriver.Firefox(options=options, service=service)

driver = webdriver.Firefox('/usr/lib/firefox/')

## Array in cui mettere i risultati
titles=[]
authors=[]
plots=[]

# Array delle pagine di qlibri da raschiare
# (La maggior parte del link è sempre uguale, ho solo memorizzato la parte saliente)
categories=["straniera","italiana"]
genres=["romanzi", "gialli", "racconti", "romanzi-storici", "fantasy", "fantascienza", "classici"]


# Per ognuna delle due categorie...
for c in categories:
    # Per ognuno dei generi...
    for g in genres:
        ## seleziono il driver giusto
        driver.get("https://www.qlibri.it/recensioni/"+g+"-narrativa-"+c+"/")    
        ##Estraggo le info (vedi funzione apposita)
        titles, authors, plots = soupScraping(driver, titles, authors, plots)

# Ho raschiato tutti i link     
# Metto i risultati in tanti file diversi
for i in range(len(plots)):
    outF = open("scraping/"+replaced(titles[i])+".txt", "w")
    outF.write(titles[i]+"\n")
    outF.write(authors[i]+"\n")
    outF.write(plots[i])
    outF.close()







mondadoriScrape.py
from selenium import webdriver
from selenium.webdriver.firefox.service import Service
from selenium.webdriver.firefox.options import Options
from bs4 import BeautifulSoup
import pandas as pd
import re


def replaced(str):
    # sostituisce alcuni caratteri di una stringa con altri.
    # serve per il passaggio da titolo/autore a parte del link
    str = str.replace(" ","-")
    str = str.replace("à","a")
    str = str.replace("è","e")
    str = str.replace("é","e")
    str = str.replace("ì","i")
    str = str.replace("ò","o")
    str = str.replace("ó","o")
    str = str.replace("ö","o")
    str = str.replace("ù","u")
    str = str.replace("ð","d")
    return str

driver = webdriver.Firefox('/usr/lib/firefox/')

## Array in cui mettere i risultati
titles=[]
authors=[]
plots=[]

# Array delle pagine di qlibri da raschiare
# (La maggior parte del link è sempre uguale, ho solo memorizzato la parte saliente)

genres=["gialli-e-thriller", "narrativa-contemporanea", "romanzi-storici-fantasy-e-fantascienza"]

# Per ognuno dei generi...
for g in genres:
    # e per ognuna delle prime 7 pagine...
    for pagenum in range(1,8):
        ## seleziono il driver giusto
        currentLink = "https://www.librimondadori.it/genere/"+g+"/page/"+str(pagenum)+"/"
        driver.get(currentLink)    
        content = driver.page_source
        soup = BeautifulSoup(content, 'html.parser')
         
        ## Estraggo le info su titolo e autore e le metto negli array appositi
        for a in soup.findAll(class_='product-info'):
            title= a.find('a', attrs={'href' : re.compile('https://www.librimondadori.it/libri/.*')})
            author=a.find('a', attrs={'href' : re.compile('https://www.librimondadori.it/autore/.*')})
            if title is None or author is None:
                break
            titles.append(title.text)
            authors.append(author.text)
        
## in differita, mi occupo delle trame  e del salvataggio in file
for quel in range(0, len(titles)):
              
    nextLink="https://www.librimondadori.it/libri/"+replaced(titles[quel])+"-"+replaced(authors[quel])+"/"
            
    driver.get(nextLink)    
    content = driver.page_source
    soup = BeautifulSoup(content, 'html.parser')

    # trovo, tra tutti i p, quelli della trama.
    # il primo è sempre il 17, seguito da un numero variabile di altri paragrafi.
    # la trama finisce quando il paragrafo successivo descrive il genere
    n=17
    plot=""
    while n<30:
        p = soup.find_all('p')[n]
        if p is None or p.text.startswith("Genere: "):
            break;
        if p.text.startswith('\n'):
            plot=""
        else:
            plot=plot+p.text+" "
            n+=1
    # piccolo controllo se la trama è nulla
    if plot is None:
        break
    plots.append(plot) 
    
    # Metto i dati del libro selezionato su un file
    # open a (new) file to write
    outF = open("scraping/"+replaced(titles[quel])+".txt", "w")
    outF.write(titles[quel]+"\n")
    outF.write(authors[quel]+"\n")
    outF.write(plot)
    outF.close()
    
# Ho raschiato tutti i link
   





genericScrape.py
from selenium import webdriver
from selenium.webdriver.firefox.service import Service
from selenium.webdriver.firefox.options import Options
from bs4 import BeautifulSoup
import pandas as pd
import re
from datetime import datetime


def replaced(str):
    # sostituisce alcuni caratteri di una stringa con altri.
    # serve per il passaggio da titolo/autore a parte del link
    str = str.replace(" ","-")
    str = str.replace("à","a")
    str = str.replace("è","e")
    str = str.replace("é","e")
    str = str.replace("ì","i")
    str = str.replace("ò","o")
    str = str.replace("ó","o")
    str = str.replace("ö","o")
    str = str.replace("ù","u")
    str = str.replace("ð","d")
    return str

## funzione che si occupa dell'output su file
def fileWrite(title, author, genre, link, plot):
    outF = open("scraping/"+replaced(title)+".txt", "w")
    outF.write(title+"\n")
    outF.write(author+"\n")
    outF.write(genre+"\n")
    outF.write(link+"\n")
    outF.write(plot)
    outF.close()

## Funzione che fa il web cscraping di titoli, autori, generi e link
#  separata dal resto per ragioni di leggibilità
def soupScraping(driver, titles, authors, genres, links, g):
    content = driver.page_source
    soup = BeautifulSoup(content, 'html.parser')

    ## Estraggo le info e le metto negli array appositi
    # non c'è bisogno di differenziare con un if sul sito, perché se soup non trova la classe giusta non avvierà neanche il for
    
    #qlibri
    for a in soup.findAll(class_='listItem'):
        title=a.find('a',href=True, attrs={'class':'jr_listingTitle'})
        author=a.find('div', attrs={'class':'fieldValue'})
        
        # piccolo controllo, se non interrompo il ciclo mi dà errore
        if title is None or author is None:
            break
        
        # Creo il link e salvo la formattazione del genere senza trattini o barre
        link = "https://www.qlibri.it/"+replaced(g)+"/"+replaced(title.text.replace("-","%11"))+"/"
        genre = g.replace("-"," ").replace("/"," ")
        
        titles.append(title.text)
        authors.append(author.text)
        genres.append(genre) 
        links.append(link)         
    #mondadorilibri 
    for a in soup.findAll(class_='product-info'):
        title= a.find('a', attrs={'href' : re.compile('https://www.librimondadori.it/libri/.*')})
        author=a.find('a', attrs={'href' : re.compile('https://www.librimondadori.it/autore/.*')})
        
        if title is None or author is None:
            break
          
        # Creo il link e salvo la formattazione del genere senza trattini o barre
        link="https://www.librimondadori.it/libri/"+replaced(title.text)+"-"+replaced(author.text)+"/"
        genre = g.replace("-"," ").replace("/"," ")
        
        titles.append(title.text)
        authors.append(author.text)
        genres.append(genre) 
        links.append(link)
            
    # Ritorno gli array per aggiornarli
    return titles, authors, genres, links

def loadingBar(X, maxX):
    # loading bar, per sapere quante trame ho scaricato in percentuale 
          
    if X > maxX:
        X = maxX
    if X < 0:
        X = 0
    percent = (int)((float)(X/maxX*100)) 
    loadState = ("#"*percent)+(" "*(100-percent))
    return str(percent) + "%	["+loadState+"]"



start_time = datetime.now()
numPlot = 1


driver = webdriver.Firefox('/usr/lib/firefox/')

# Array delle pagine da raschiare
# (La maggior parte del link è sempre uguale, ho solo memorizzato la parte saliente: 
# in librimondadori cambia solo il genere, in qlibri genere e categoria)

mGenres = ["gialli-e-thriller", "narrativa-contemporanea", "romanzi-storici-fantasy-e-fantascienza"] # generi di mondadorilibri
qGenres = ["romanzi", "gialli", "racconti", "romanzi-storici", "fantasy", "fantascienza", "classici"] # generi di qlibri
qCategories = ["straniera","italiana"]

# per ognuno dei due siti...
for site in ["qlibri", "mondadorilibri", ]:
    ### Per cominciare, definisco/vuoto gli array in cui mettere i risultati
    titles = []
    authors = []
    genres = []
    links = []
    plots = []
    
    ### In base al sito, trovo i link alle pagine generiche con le liste di titoli e autori
    genericLinks = []
    if site == "qlibri":
        for c in qCategories:
            for g in qGenres:
                ## seleziono il driver giusto
                genericLinks.append("https://www.qlibri.it/recensioni/"+g+"-narrativa-"+c+"/") 
    elif site == "mondadorilibri": 
        for g in mGenres:
            for pagenum in range(1,8): # numero di pagine
                genericLinks.append("https://www.librimondadori.it/genere/"+g+"/page/"+str(pagenum)+"/")
                print("https://www.librimondadori.it/genere/"+g+"/page/"+str(pagenum)+"/")
    
    ### Titolo, autore, genere, link
    # per ogni link generico trovato...
    for nextLink in genericLinks:
        driver.get(nextLink)
        ## Trovo il genere a cui mi sto riferendo
        g = nextLink.split("/")[4]
        if site == "qlibri": # qlibri ha bisogno di ulteriore formattazione
            splittedg= g.split("-")
            g = splittedg[len(splittedg)-2]+"-"+splittedg[len(splittedg)-1]+"/" # prima la categoria, poi una barra... 
            for i in range (len(splittedg)-2):
                g=g+splittedg[i]+"-" #poi il genere
            g = g[:-1] #tolgo l'ultimo trattino
        ##Estraggo titolo, autore e link alla trama estesa (vedi funzione apposita)
        titles, authors, genres, links = soupScraping(driver, titles, authors, genres, links, g)
    
    ### Trama
    for plotLink in links:
        driver.get(plotLink)    
        content = driver.page_source
        soup = BeautifulSoup(content, 'html.parser')
        if site == "qlibri":
            # Qlibri è più simpatico: ha direttamente una classe per la trama
            p = soup.find('div', attrs={'class':'contentFulltext'})
            plot = p.text[4:] # via quei brutti tab dall'inizio della trama!
            if plot is None:
                break
        elif site == "mondadorilibri":
            # Mondadorilibri, invece, definisce tutti i contenuti della pagina con la p di paragrafo.
            # Devo trovare, tra tutti i p, quelli della trama.
            # il primo è sempre il 17 (prima ci sono info varie), seguito da un numero variabile di altri paragrafi.
            # la trama finisce quando il paragrafo successivo descrive il genere
            n=17
            plot=""
            while n<30:
                p = soup.find_all('p')[n]
                if p is None or p.text.startswith("Genere: "): # Se la trama è nulla o se si è già passati al genere
                    break;
                else:
                    plot=plot+p.text+" "
                    n+=1
            plot.strip('\n')
            # piccolo controllo se la trama è nulla
            if plot is None:
                break
        
        # Una serie di print per dare un'idea della situazione
        # (questo programma ci mette 25 minuti a girare, diamogli almeno una barra di caricamento!)
        d = datetime.now()-start_time
        
        print("\nTrama n. "+str(numPlot)+"\tTempo passato= "+ str(d.seconds // 3600)+":"+str(d.seconds % 3600 // 60)+":"+str(d.seconds%60))
        print("qlibri:		"+loadingBar(numPlot, max(len(titles), 663)))
        print("mondadorilibri:	"+loadingBar(numPlot-663, min(len(titles), 245)))
        numPlot += 1

        plots.append(plot)
                    

    ### Output su file
    for i in range (0, len(titles)):
        fileWrite(titles[i], authors[i], genres[i], links[i], plots[i])

